KOGITO-7766 - Add Data Index Addon example using PostgreSQL persistence (#1532)
kiegroup/kogito-examples: serverless-workflow-examples/serverless-workflow-data-index-quarkus/README.md

# Kogito Serverless Workflow - Data Index Example

## Description

This example contains a simple workflow service that demonstrates how to use Data Index Addon as part of the Kogito runtime. 

> A callback is a state that invokes an action and wait for an event (event that will be eventually fired by the external service notified by the action), so this example needs an event broker.

Reviewer: ```suggestion
A callback is a state that invokes an action and waits for an event (an event that will be eventually fired by the external service notified by the action), so this example needs an event broker.
```

> This example consist of a callback state that waits for an event arriving on wait channel. Its action publish an event on resume channel. The event published on resume channnel is modified and republished into the wait channel by `PrintService`, which simulates an external service. 

Reviewer: ```suggestion
This example consists of a callback state that waits for an event arriving on the wait channel. Its action is to publish an event on resume channel. The event published on resume channel is modified and republished into the wait channel by `PrintService`, which simulates an external service. 
```

The service is described using JSON format as defined in the 
[CNCF Serverless Workflow specification](https://github.com/serverlessworkflow/specification).

## Infrastructure requirements

### Kafka

This quickstart requires an Apache Kafka to be available and by default expects it to be on default port and localhost.

* Install and Startup Kafka Server / Zookeeper

https://kafka.apache.org/quickstart

> To publish and consume the event, topic "move" is used. 

Reviewer 2: ```suggestion
To publish and consume the event, the topic `move` is used. 
```

Optionally and for convenience, a docker-compose [configuration file](docker-compose/docker-compose.yml) is
provided in the path [docker-compose/](docker-compose/), where you can just run the command from there:

```sh
docker-compose up
```

In this way a container for Kafka will be started on port 9092.

### PostgreSQL

This example also requires persistence with a PostgreSQL server.

> Configuration for setting up the connection can be found in [applications.properties](src/main/resources/application.properties) file, which

Reviewer 2: ```suggestion
The configuration for setting up the connection can be found in [applications.properties](src/main/resources/application.properties) file, which
```

follows the Quarkus JDBC settings, for more information please check [JDBC Configuration Reference](https://quarkus.io/guides/datasource#jdbc-configuration).

Optionally and for convenience, a docker-compose [configuration file](docker-compose/docker-compose.yml) is
provided in the path [docker-compose/](docker-compose/), where you can just run the command from there:

```sh
docker-compose up
```

In this way a container for PostgreSQL will be started on port 5432.

## Installing and Running

### Prerequisites
 
You will need:
  - Java 11+ installed
  - Environment variable JAVA_HOME set accordingly
  - Maven 3.8.1+ installed
  - Docker and Docker Compose to run the required example infrastructure.

When using native image compilation, you will also need: 

>     - GraalVM 20.3+ installed

Reviewer 2: ```suggestion
    - GraalVM 22.2+ installed
```

    - Environment variable GRAALVM_HOME set accordingly
    - GraalVM native image needs as well native-image extension: https://www.graalvm.org/reference-manual/native-image/
    - Note that GraalVM native image compilation typically requires other packages (glibc-devel, zlib-devel and gcc) to be installed too, please refer to GraalVM installation documentation for more details.

NOTE: Quarkus provides a way of creating a native Linux executable without GraalVM installed, leveraging a container runtime such as Docker or Podman. More details in  https://quarkus.io/guides/building-native-image#container-runtime 

### Compile and Run in Local Dev Mode

```sh
mvn clean package quarkus:dev
```

### Start infrastructure services

> You should start all the services before you execute any of the **Data Index** example, to do that please execute:

Reviewer 2: ```suggestion
You should start all the services before you execute any of the **Data Index** example. To do that please execute:
```

```sh
mvn clean package -Dcontainer
```

For Linux and MacOS:

1. Open a Terminal
2. Go to docker-compose folder
3. Run ```docker-compose up```

```bash
cd docker-compose && docker-compose up
```

> Tip: If you get permission denied while creating the postgres container, consider using SELinux context.

Reviewer 2: ```suggestion
TIP: If you get a `permission denied` error while creating the postgresql container, consider using SELinux context.
```

> Tip: If you get permission denied while creating the postgres container, consider to use SELinux context.

Reviewer: ```suggestion
Tip: If you get permission denied while creating the PostgreSQL container, consider using SELinux context.
```

Update the following line:
```yaml
    - ./sql:/docker-entrypoint-initdb.d
```
to
```yaml
    - ./sql:/docker-entrypoint-initdb.d:Z
```

Once all services bootstrap, the following ports will be assigned on your local machine:

- PostgreSQL: 5432
- Kafka: 9092
- PgAdmin: 8055

> - serverless-workflow-service :8080

Reviewer 2: ```suggestion
- serverless-workflow-service: 8080
```

> **_NOTE:_**  This step requires the project to be compiled, please consider running a ```mvn clean package -Dcontainer``` command on the project root before running the ```docker-compose up``` for the first time or any time you modify the project.

Once started you can simply stop all services by executing the ```docker-compose -f docker-compose.yml stop```.

All created containers can be removed by executing the ```docker-compose -f docker-compose.yml rm```.

### Compile and Run using Local Native Image
Note that this requires GRAALVM_HOME to point to a valid GraalVM installation

```sh

> mvn clean package -Pnative

Reviewer 2: ```suggestion
mvn clean package -Dnative
```
